script/tool/lib/src/common/repository_package.dart

```diff
@@ -214,4 +232,180 @@ class RepositoryPackage {
+        }
+      } else {
+        throw FormatException(
+            'Found non-YAML file in pending_changelogs: ${file.path}');
+      }
+    }
+
+    for (final File file in pendingChangelogFiles) {
+      try {
+        entries
+            .add(PendingChangelogEntry._parse(file.readAsStringSync(), file));
+      } on FormatException catch (e) {
+        throw FormatException(
+            'Malformed pending changelog file: ${file.path}\n$e');
+      }
+    }
+    return entries;
+  }
+}
+
+/// A class representing the parsed content of a `ci_config.yaml` file.
+class CiConfig {
+  /// Creates a [CiConfig] from a parsed YAML map.
+  CiConfig._(this.isBatchRelease);
+
+  /// Parses a [CiConfig] from a YAML string.
+  ///
+  /// Throws if the YAML is not a valid ci_config.yaml.
+  factory CiConfig._parse(String yaml) {
+    final Object? loaded = loadYaml(yaml);
+    if (loaded is! YamlMap) {
+      throw const FormatException('Root of ci_config.yaml must be a map.');
+    }
+
+    _checkCiConfigEntries(loaded, syntax: _validCiConfigSyntax);
+
+    bool isBatchRelease = false;
+    final Object? release = loaded['release'];
+    if (release is Map) {
+      isBatchRelease = release['batch'] == true;
+    }
+
+    return CiConfig._(isBatchRelease);
+  }
+
+  static const Map<String, Object?> _validCiConfigSyntax = <String, Object?>{
+    'release': <String, Object?>{
+      'batch': <bool>{true, false}
+    },
+  };
+
+  /// Returns true if the package is configured for batch release.
+  final bool isBatchRelease;
+
+  static void _checkCiConfigEntries(YamlMap config,
+      {required Map<String, Object?> syntax, String configPrefix = ''}) {
+    for (final MapEntry<Object?, Object?> entry in config.entries) {
+      if (!syntax.containsKey(entry.key)) {
+        throw FormatException(
+            'Unknown key `${entry.key}` in config${_formatConfigPrefix(configPrefix)}, the possible keys are ${syntax.keys.toList()}');
+      } else {
+        final Object syntaxValue = syntax[entry.key]!;
+        final String newConfigPrefix = configPrefix.isEmpty
+            ? entry.key! as String
+            : '$configPrefix.${entry.key}';
+        if (syntaxValue is Set) {
+          if (!syntaxValue.contains(entry.value)) {
+            throw FormatException(
+                'Invalid value `${entry.value}` for key${_formatConfigPrefix(newConfigPrefix)}, the possible values are ${syntaxValue.toList()}');
+          }
+        } else if (entry.value is! YamlMap) {
+          throw FormatException(
+              'Invalid value `${entry.value}` for key${_formatConfigPrefix(newConfigPrefix)}, the value must be a map');
+        } else {
+          _checkCiConfigEntries(entry.value! as YamlMap,
+              syntax: syntaxValue as Map<String, Object?>,
+              configPrefix: newConfigPrefix);
+        }
+      }
+    }
+  }
+
+  static String _formatConfigPrefix(String configPrefix) =>
+      configPrefix.isEmpty ? '' : ' `$configPrefix`';
+}
+
```

Comment:
Let's move everything from here on to a separate helper file next to this one, rather than dumping all the pending changelog logic into `repository_package.dart`.

That file should also start with a file-level comment explaining how the pending changelog system works at a high level (the directory, example entry files, etc.)